### PR TITLE
Enable `pip` caching

### LINKED ISSUE #3623: Improve robustness against Python build failures on pull requests due to network flakiness
https://grpc-testing.appspot.com/job/gRPC_pull_requests/3484/

### PR DESCRIPTION
**Pull request by soltanmm**
@nathanielmanistaatgoogle I threw in a little clean-up refactoring in build_python.sh in prep for whatever comes of future efforts to split external dependency build steps from gRPC build steps.

Somewhat fixes #3623 (but there'll be more on that front for everyone's general build-flakiness-due-to-network-snafus). The basic requirement, as with any caching solution, is that at least one of the previous builds with the same version requirements succeeded.

Tagged DNM to watch what happens on Jenkins.

### CODE REVIEW

File: tools/run_tests/build_python.sh [hunk 1]
```diff
@@ -39,6 +39,46 @@ GRPCIO=$ROOT/src/python/grpcio
 GRPCIO_TEST=$ROOT/src/python/grpcio_test
 GRPCIO_HEALTH_CHECKING=$ROOT/src/python/grpcio_health_checking
 
+#
```
Comment (nathanielmanistaatgoogle) [inline tools/run_tests/build_python.sh]: In languages that aren't Bash when I see stanza-delineating comments like these I suggest to the author that the file has gotten too large and should be broken up. Is that the case here? This stanza is just two functions; why bother with the comments anyway?

File: tools/run_tests/build_python.sh [hunk 2]
```diff
@@ -39,6 +39,46 @@ GRPCIO=$ROOT/src/python/grpcio
...
+install_grpcio_health_checking() {
+  pip install $GRPCIO_HEALTH_CHECKING
+}
+#
+# End install steps.
+#
+
+# Cleans the environment of previous installations
+clean_grpcio_all() {
+  (yes | pip uninstall grpcio) || true
```
Comment (nathanielmanistaatgoogle) [inline tools/run_tests/build_python.sh]: "pip uninstall --yes" strikes me as a less drastic means to the same end as "yes | pip uninstall". Is it preferable or are my tastes and sensibilities missing something?

File: tools/run_tests/build_python.sh [hunk 3]
```diff
@@ -39,6 +39,46 @@ GRPCIO=$ROOT/src/python/grpcio
...
+  (yes | pip uninstall grpcio_test) || true
```
Comment (nathanielmanistaatgoogle) [inline tools/run_tests/build_python.sh]: Is `|| true` swallowing potentially important failure/error information?

Comment (soltanmm) [inline tools/run_tests/build_python.sh]: Nope. It's only there to swallow errors on the possibility that the testing user ctrl-c'd the script mid-build and we need to realize that our testing environment is corrupted at some later point. It isn't so much a 'clean' as an 'ensure as clean as we can'.

File: tools/run_tests/build_python.sh (hunk 2 above)
Comment (soltanmm) [inline tools/run_tests/build_python.sh]: I'd usually say whichever, but right now the pipes lining up without extra spacing has a particular aesthetic quality and the command has a more positive zen to it than with a bunch of minus signs everywhere. I'm hesitant to disturb the feng shui. :-P

File: tools/run_tests/build_python.sh (hunk 1 above)
Comment (soltanmm) [inline tools/run_tests/build_python.sh]: Eh, no reason. Got rid of 'em.

File: tools/jenkins/grpc_jenkins_slave/Dockerfile [hunk 4]
```diff
@@ -126,10 +126,11 @@ RUN /bin/bash -l -c "gem install bundler --no-ri --no-rdoc"
...
     python-all-dev \
     python3-all-dev \
-    python-pip \
-    python-virtualenv
+    python-pip
 
 # Install Python packages from PyPI
+RUN pip install pip --upgrade
+RUN pip install virtualenv
 RUN pip install futures==2.2.0 enum34==1.0.4 protobuf==3.0.0a2
```
Comment (nathanielmanistaatgoogle) [inline tools/jenkins/grpc_jenkins_slave/Dockerfile]: Elsewhere in this pull request I see "pip install -r requirements.txt"; is that an option here? (#1564)

**Comment by nathanielmanistaatgoogle** [comment]
LGTM but @nicolasnoble will lead review.

File: tools/jenkins/grpc_jenkins_slave/Dockerfile (hunk 4 above)
Comment (soltanmm) [inline tools/jenkins/grpc_jenkins_slave/Dockerfile]: The gRPC repository doesn't exist at this point. I'm actually not entirely certain what the point of this line is...

**Comment by soltanmm** [comment]
(confirming the existence of the populated `pip` caches on the testing servers outside of the images)

**Pull request merged**